File: test-jasmine/linked-list.js
'use strict'

var LinkedListNode = require('../index').LinkedListNode
var LinkedList = require('../index').LinkedList


describe('LinkedListNode', function() {
    it('can update its next node consistently', function() {
        var nodeA = new LinkedListNode('a')
        var nodeB = new LinkedListNode('b')
        var nodeC = new LinkedListNode('c')
        nodeA.next = nodeB
        expect(nodeA.next).toBe(nodeB)
        expect(nodeB.prev).toBe(nodeA)
        nodeA.next = nodeC
        expect(nodeA.next).toBe(nodeC)
        expect(nodeC.prev).toBe(nodeA)
        expect(nodeB.prev).toBeFalsy()
    })

    it('can update its previous node consistently', function() {
        var nodeA = new LinkedListNode('a')
        var nodeB = new LinkedListNode('b')
        var nodeC = new LinkedListNode('c')
        nodeA.prev = nodeB
        expect(nodeA.prev).toBe(nodeB)
        expect(nodeB.next).toBe(nodeA)
        nodeA.prev = nodeC
        expect(nodeA.prev).toBe(nodeC)
        expect(nodeC.next).toBe(nodeA)
        expect(nodeB.next).toBeFalsy()
    })
})

describe('LinkedList', function() {
    var list
    beforeEach(function() {
        list = new LinkedList()
    })

    it('addFirst, getFirst, getLast', function() {
        list.addFirst(1)
        expect(list.getFirst()).toBe(1)
        expect(list.getLast()).toBe(1)
    })

    it('addLast, getFirst, getLast', function() {
        list.addLast(5)
        expect(list.getFirst()).toBe(5)
        expect(list.getLast()).toBe(5)
    })

    it('addFirst twice, getFirst, getLast', function() {
        list.addFirst(1)
        list.addFirst(2)
        expect(list.getFirst()).toBe(2)
        expect(list.getLast()).toBe(1)
    })

    it('addFirst returns new node with given element', function() {
        expect(list.addFirst(-3).value).toBe(-3)
    })

    it('can remove a node from the middle of a list', function() {
        var nodeA = list.addLast(1)
        var nodeB = list.addLast(2)
        var nodeC = list.addLast(3)
        nodeB.remove()
        expect(nodeA.next).toBe(nodeC)
        expect(nodeC.prev).toBe(nodeA)
    })

    it('can remove the only node from a list', function() {
        list.addFirst(true).remove()
        expect(list.first).toBeFalsy()
        expect(list.last).toBeFalsy()
        list.addLast(false).remove()
        expect(list.first).toBeFalsy()
        expect(list.last).toBeFalsy()
    })

    it('removes one of two nodes, making other node both first and last', function() {
        var nodeA = list.addFirst(123)
        var nodeB = list.addLast(456)
        nodeB.remove()
        expect(list.getFirstNode()).toBe(nodeA)
        expect(list.getLastNode()).toBe(nodeA)
        nodeB = list.addFirst(789)
        nodeB.remove()
        expect(list.getFirstNode()).toBe(nodeA)
        expect(list.getLastNode()).toBe(nodeA)
    })

    it('removes a node twice safely', function() {
        var node = list.addFirst({})
        node.remove()
        node.remove()
        expect(list.empty).toBe(true)
    })

    it('can clear a list', function() {
        list.addLast(1)
        list.addLast(2)
        list.addLast(3)
        list.clear()
        expect(list.empty).toBe(true)
    })

    it('removes nodes without breaking iteration', function() {
        var callback = jasmine.createSpy('callback')

        list.forEach(callback)
        expect(callback.calls.all()).toEqual([])
        callback.calls.reset()

        list.addLast(1)
        var node = list.addLast(2)
        list.addLast(3)
        list.forEach(callback)
        expect(callback.calls.allArgs()).toEqual([
            [1], [2], [3]
        ])
        callback.calls.reset()

        node.remove()
        list.forEach(callback)
        expect(callback.calls.allArgs()).toEqual([
            [1], [3]
        ])
        callback.calls.reset()

        node.remove()
        list.forEach(callback)
        expect(callback.calls.allArgs()).toEqual([
            [1], [3]
        ])
        callback.calls.reset()
    })

    it('removes nodes during iteration', function() {
        var nodeA = list.addLast(1)
        var nodeB = list.addLast(2)
        var nodeC = list.addLast(3)

        var callback = jasmine.createSpy('callback').and.callFake(function(node) {
            node.remove()
        })
        list.forEachNode(callback)
        expect(callback.calls.allArgs()).toEqual([
            [nodeA], [nodeB], [nodeC]
        ])

        expect(list.head.next).toBe(list.tail)
        expect(list.tail.prev).toBe(list.head)
    })

    it('removes consecutive nodes during iteration', function() {
        var nodeA = list.addLast(1)
        var nodeB = list.addLast(2)
        var nodeC = list.addLast(3)

        var called = false
        list.forEachNode(function(node) {
            if(node === nodeA) {
                expect(called).toBe(false)
                called = true
                nodeA.remove()
                nodeB.remove()
            }
            else {
                expect(called).toBe(true)
                expect(node).toBe(nodeC)
            }
        })
        expect(list.head.next).toBe(nodeC)
        expect(list.tail.prev).toBe(nodeC)
    })

})

File: src/esquire/LinkedList.kt
package esquire

import java.util.*


open class LinkedListNode<TElement>(val value: TElement): Registration {
    var removed = false

    private var _prev: LinkedListNode<TElement>? = null
    var prev: LinkedListNode<TElement>?
        get() = _prev
        set(other) {
            if(_prev != null)
                _prev!!._next = null
            _prev = other
            if(_prev != null) {
                if(_prev!!._next != null)
                    _prev!!._next!!._prev = null
                _prev!!._next = this
            }
        }

    private var _next: LinkedListNode<TElement>? = null
    var next: LinkedListNode<TElement>?
        get() = _next
        set(other) {
            if(_next != null)
                _next!!._prev = null
            _next = other
            if(_next != null) {
                if(_next!!._prev != null)
                    _next!!._prev!!._next = null
                _next!!._prev = this
            }
        }

    fun insertValueAfter(element: TElement): LinkedListNode<TElement> {
        val node = LinkedListNode(element)
        insertAfter(node)
        return node
    }

    fun insertValueBefore(element: TElement): LinkedListNode<TElement> {
        val node = LinkedListNode(element)
        insertBefore(node)
        return node
    }

    fun insertAfter(node: LinkedListNode<TElement>) {
        node.next = next
        node.prev = this
    }

    fun insertBefore(node: LinkedListNode<TElement>) {
        node.prev = prev
        node.next = this
    }

    override fun removeImpl() {
        if(removed)
            return

        val oldPrev = prev
        val oldNext = next
        removed = true
        if(oldPrev != null)
            oldPrev._next = oldNext
        if(oldNext != null)
            oldNext._prev = oldPrev
    }

    override var untilRegistrations: ArrayList<Registration>? = null
}

class LinkedList<TElement> {
    val head = LinkedListNode<TElement>(null)
    val tail = LinkedListNode<TElement>(null)

    init {
        head.next = tail
    }

    fun addFirst(element: TElement) = head.insertValueAfter(element)
    fun addLast(element: TElement) = tail.insertValueBefore(element)

    fun getFirst() = getFirstNode().value
    fun getLast() = getLastNode().value

    fun getFirstNode(): LinkedListNode<TElement> {
        if(empty)
            throw NoSuchElementException()
        return head.next!!
    }

    fun getLastNode(): LinkedListNode<TElement> {
        if(empty)
            throw NoSuchElementException()
        return tail.prev!!
    }

    val empty: Boolean
        get() = head.next === tail

    fun clear() {
        head.next = tail
    }

    fun forEach(callback: (TElement)->Unit) = forEachNode { callback(it.value) }

    fun forEachNode(callback: (LinkedListNode<TElement>)->Unit) {
        var node = head
        while(true) {
            node = node.next!!
            if(node == tail)
                return
            if(node.removed)
                continue
            callback(node)
        }
    }
}
